projects/meta_cl/networks/dendritic_networks.py

```diff
@@ -184,6 +182,116 @@ def conv_block(cls, in_channels, out_channels, kernel_size, stride, padding):
+
+        inds = dendrite_activations.abs().max(dim=2).indices
+        inds = inds.unsqueeze(dim=2)
+        dendrite_activations = torch.gather(dendrite_activations, dim=2, index=inds)
+        dendrite_activations = dendrite_activations.squeeze(dim=2)
+        dendrite_activations = torch.sigmoid(dendrite_activations)
+        return y * dendrite_activations
+
+    def forward(self, x):
+
+        mod = self.modulation(x)
+        pred = self.prediction(x)
+
+        dendrite_activations = self.segments(mod)
+        out = self.apply_dendrites(pred, dendrite_activations)
+
+        out = self.classifier(out)
+
+        return out
+
+
+class CloserToANMLDendriticNetwork(nn.Module):
+    """
+    An iteration of a dendritic network, based closely on `ANML`_. (The conv + max-pool
+    layers are identical). It's similar to ANMLDendriticNetwork, but does not include a
+    linear layer atop the modulation network. This makes it more similar to ANML's
+    original network. In fact, with `num_segments=1, dendrite_sparsity=0`, they should
+    be identical.
+
+    .. _ANML: nupic.research/projects/meta_cl/networks/anml_networks.py
+
+    Input shape expected is 3 x 28 x 28 (C x W x H) but an adaptive average pooling
+    layer helps accept other heights and width.
+
+    With default parameters and `num_classes-963`, it uses 5,132,832 weights-on
+    out of a total of 13,960,323 weights. In comparison, ANML uses 5,963,139 weights
+    and OML uses 5,172,675 weights. Thus, these is still room to add more.
+
+    The default parameters are chosen so that the number of on weights for this network
+    is close to that of ANMLDendriticNetwork.
+    """
+
+    def __init__(self, num_classes,
+                 num_segments=10,
+                 dendrite_sparsity=0.856,
+                 dendrite_bias=None):
+
+        super().__init__()
+
+        self.segments = DendriteSegments(
+            num_units=2304,
+            num_segments=num_segments,
+            dim_context=1008,
+            sparsity=dendrite_sparsity,
+            bias=dendrite_bias,
+        )
+        self.classifier = nn.Linear(2304, num_classes)
+
+        self.prediction = nn.Sequential(
+            *self.conv_block(3, 256, 3, 1, 0),
+            nn.MaxPool2d(kernel_size=2, stride=2),
+            *self.conv_block(256, 256, 3, 1, 0),
+            nn.MaxPool2d(kernel_size=2, stride=2),
+            *self.conv_block(256, 256, 3, 1, 0),
+            nn.AdaptiveAvgPool2d(output_size=(3, 3)),
+            nn.Flatten(),
+        )
+
+        self.modulation = nn.Sequential(
+            *self.conv_block(3, 112, 3, 1, 0),
+            nn.MaxPool2d(kernel_size=2, stride=2),
+            *self.conv_block(112, 112, 3, 1, 0),
+            nn.MaxPool2d(kernel_size=2, stride=2),
+            *self.conv_block(112, 112, 3, 1, 0),
+            nn.AdaptiveAvgPool2d(output_size=(3, 3)),
+            nn.Flatten(),
+        )
+
+        # Apply Kaiming initialization
+        self.reset_params()
```

Comment:
Not sure this is necessary. I believe Kaiming is the default initialization for Linear and Conv2d. (https://discuss.pytorch.org/t/whats-the-default-initialization-methods-for-layers/3157/21)